lib/Parse/ASTGen.cpp

```diff
@@ -10,80 +10,738 @@
+  TypeRepr *WrappedType = generate(Type.getWrappedType(), Loc);
+  auto QuestionLoc = advanceLocBegin(Loc, Type.getQuestionMark());
+  return new (Context) OptionalTypeRepr(WrappedType, QuestionLoc);
+}
+
+TypeRepr *ASTGen::generate(ImplicitlyUnwrappedOptionalTypeSyntax Type,
+                           SourceLoc &Loc) {
+  TypeRepr *WrappedType = generate(Type.getWrappedType(), Loc);
+  auto ExclamationLoc = advanceLocBegin(Loc, Type.getExclamationMark());
+  return new (Context)
+      ImplicitlyUnwrappedOptionalTypeRepr(WrappedType, ExclamationLoc);
+}
+
+TypeRepr *ASTGen::generate(UnknownTypeSyntax Type, SourceLoc &Loc) {
+  auto ChildrenCount = Type.getNumChildren();
+
+  if (ChildrenCount == 3 || ChildrenCount == 4) {
+    auto Element = Type.getChild(0)->getAs<TypeSyntax>();
+    auto LSquare = Type.getChild(1)->getAs<TokenSyntax>();
+    auto Last = Type.getChild(ChildrenCount - 1);
+
+    if (Element && LSquare && LSquare->getTokenKind() == tok::l_square) {
+      auto ElementType = generate(*Element, Loc);
+      auto Begin = advanceLocBegin(Loc, *Element);
+      auto End = advanceLocBegin(Loc, *Last);
+      return new (Context) ArrayTypeRepr(ElementType, {Begin, End});
+    }
+  }
+
+  if (ChildrenCount == 2) {
+    auto Element = Type.getChild(0)->getAs<TypeSyntax>();
+    auto LSquare = Type.getChild(1)->getAs<TokenSyntax>();
+
+    if (Element && LSquare && LSquare->getTokenKind() == tok::l_square) {
+      return generate(*Element, Loc);
+    }
+  }
+
+  if (ChildrenCount >= 2) {
+    auto Protocol = Type.getChild(0)->getAs<TokenSyntax>();
+
+    if (Protocol && Protocol->getText() == "protocol") {
+      auto LAngle = Type.getChild(1);
+
+      SmallVector<TypeSyntax, 4> Protocols;
+      for (unsigned i = 2; i < Type.getNumChildren(); i++)
+        if (auto PType = Type.getChild(i)->getAs<TypeSyntax>())
+          Protocols.push_back(*PType);
+
+      auto RAngle = Type.getChild(ChildrenCount - 1);
+
+      auto ProtocolLoc = advanceLocBegin(Loc, *Protocol);
+      auto LAngleLoc = advanceLocBegin(Loc, *LAngle);
+      auto RAngleLoc = advanceLocBegin(Loc, *RAngle);
+
+      SmallVector<TypeRepr *, 4> ProtocolTypes;
+      for (auto &&P : llvm::reverse(Protocols))
+        ProtocolTypes.push_back(generate(P, Loc));
+      std::reverse(std::begin(ProtocolTypes), std::end(ProtocolTypes));
+
+      return CompositionTypeRepr::create(Context, ProtocolTypes, ProtocolLoc,
+                                         {LAngleLoc, RAngleLoc});
+    }
+  }
+
+  if (ChildrenCount == 1) {
+    auto Keyword = Type.getChild(0)->getAs<TokenSyntax>();
+
+    auto IsAtStartOfLine = [](const TokenSyntax Token) {
+      auto LeadingTrivia = Token.getLeadingTrivia();
+      if (LeadingTrivia.empty())
+        return false;
+      auto LastLeadingTriviaKind = Token.getLeadingTrivia().back().getKind();
+      return LastLeadingTriviaKind == TriviaKind::Newline ||
+             LastLeadingTriviaKind == TriviaKind::CarriageReturnLineFeed;
```

Comment:
This `IsAtStartOfLine` is not quite correct.
```
let abc: /*
*/ class
```
This `class` is "at start of line" but the trivia doesn't contain 'Newline' or `CarriageReturnLineFeed` piece.
Also, can't we just omit this `IsAtStartOfLine` check? The parser already checked `!isAtStartOfLine()`, so if we reach here, I think we can just construct `ErrorTypeRepr` if it's keyword token.